### ISSUE TITLE
A strange processing methods visibility

### ISSUE THREAD

**Issue by bozaro** [rust-lang/rust#22684]
rustc 1.0.0-nightly (2b01a37ec 2015-02-21) (built 2015-02-22)

I create file with content:
```
use std::hash::SipHasher;
use std::hash::Hasher;

fn main() {
}

fn compile_fail() {
	let hasher = SipHasher::new();
	hasher.write(&[0, 1, 2]);
}

fn compile_work() {
	let sip_hasher = SipHasher::new();
	let hasher: &mut Hasher = &mut sip_hasher;
	hasher.write(&[0, 1, 2]);
}
```

And have compilation errors:
```
private.rs:9:2: 9:26 error: method `write` is private
private.rs:9 	hasher.write(&[0, 1, 2]);
             	^~~~~~~~~~~~~~~~~~~~~~~~
error: aborting due to previous error
```

I do not understand what is happening due to this error, because this cast is essentially doing nothing.

**Comment by edwardw** [comment]
`SipHasher` has both a private `write` method and a public one implemented in `Hasher` trait for it. The former shadows the later. Probably related to #12808.

**Comment by edwardw** [comment]
This is rather confusing from user's point of view. Because by definition, a private method doesn't show up in the rustdoc. So looking up rustdoc doesn't help diagnosing this particular error.

I wonder what is the optimal solution here. Maybe an help message of some sort?

**Comment by steveklabnik** [comment]
I've created a self-contianed example of this behavior:

```
mod foo {
    pub struct Foo;

    impl Foo {
        fn bar(&self) {
            println!("private bar in foo");
        }
    }

    pub trait Baz {
        fn bar(&self);
    }

    impl Baz for Foo {
        fn bar(&self) {
            println!("public bar in foo from Baz");
        }
    }
}

fn main() {
    let mut f = foo::Foo;
    f.bar(); // doesn't work

    let b: &mut foo::Baz = &mut f;

    b.bar(); // works
}
```

/cc @rust-lang/lang . I am not sure that this is fixible? It would seem like a big change to modify these rules.

**Comment by aturon** [comment]
Nominating for discussion at next lang team meeting.

**Comment by nikomatsakis** [comment]
As @edwardw said, this is (to my mind) the same "FIXME" as #12808 (which just talks about fields). I think we can probably change this, because we'd just be making more code compile, but I've not thought super hard about that.

**Comment by nikomatsakis** [comment]
In @rust-lang/lang meeting we decided that this should indeed follow the same lookup rules as #12808.

**Comment by nikomatsakis** [comment]
triage: P-medium

**Comment by jseyfried** [comment]
@nikomatsakis I fixed this in #31938.
